### DISCUSSION TITLE
UPRN

### DISCUSSION THREAD

**Discussion by MrMikeRose** [digital-land/data-standards-backlog#72] [Planning consideration]
UPRN chat - go on - join in!

**Comment by colmjude** [comment]
## Existing dataset

There is an existing [UPRN dataset made available by OS](https://www.ordnancesurvey.co.uk/products/os-open-uprn).

**Reply by colmjude** [reply]
## What we found

### Folder Contents:

The following files were found in the folder:

- licence.txt: Size: 192 bytes
- osopenuprn_202409.gpkg: Size: 7043998720 bytes
- versions.txt: Size: 88 bytes

Total Folder Size: 7043999000 bytes

### Data summary:

Geometry Type: Point
Number of Records: 40909714
Features/Attributes: UPRN, X_COORDINATE, Y_COORDINATE, LATITUDE, LONGITUDE, geometry
Coordinate Reference System (CRS): EPSG:27700
File Size: 7043998720 bytes

**UPRN has a digit range of 1 to 12 digits.**

### Examples

Geojson example:

```
{
  "type": "FeatureCollection",
  "name": "uprn_sample",
  "crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:EPSG::27700" } },
  "features": [
    { 
      "type": "Feature", 
      "properties": { 
        "UPRN": 1, 
        "X_COORDINATE": 358260.66, 
        "Y_COORDINATE": 172796.58, 
        "LATITUDE": 51.4526015, 
        "LONGITUDE": -2.602075 
      }, 
      "geometry": { "type": "Point", "coordinates": [ 358260.659999999974389, 172796.579999999987194 ] } 
    }
  ]
}
```

An example of a UPRN from each digital length (1-12):

UPRN | X_COORDINATE | Y_COORDINATE | LATITUDE | LONGITUDE | geometry
-- | -- | -- | -- | -- | --
1 | 358260.66 | 172796.58 | 51.4526015 | -2.602075 | POINT (358260.66 172796.58)
26 | 352967 | 181077 | 51.5266333 | -2.6793612 | POINT (352967 181077)
100 | 358882 | 179797 | 51.5155886 | -2.5939513 | POINT (358882 179797)
1000 | 358574 | 179525 | 51.5131204 | -2.5983575 | POINT (358574 179525)
10000 | 358794 | 178238 | 51.501565 | -2.5950364 | POINT (358794 178238)
100000 | 357703 | 173226 | 51.4564211 | -2.6101518 | POINT (357703 173226)
1108655 | 527098 | 408100 | 53.5541715 | -0.082908 | POINT (527098 408100)
10089436 | 528208 | 409128 | 53.5631368 | -0.0657418 | POINT (528208 409128)
100000001 | 548133 | 186061 | 51.5539936 | 0.1352971 | POINT (548133 186061)
1003388526 | 546039 | 211570 | 51.783745 | 0.1157692 | POINT (546039 211570)
10000000102 | 519650 | 151523 | 51.2503325 | -0.2869582 | POINT (519650 151523)
100000000001 | 414650 | 558764 | 54.923363 | -1.7729638 | POINT (414650 558764)

**Reply by colmjude** [reply]
This dataset does not seem to include "historic" UPRNs.
Therefore, we assume they are removed from the dataset published by OS. This is an assumption we'll need to check with OS and by comparing future versions of the dataset.

Here is a bit of background on the UPRN process and when UPRNs are marked as historic.
![miro-diagramming-lifecycle-UPRN](https://github.com/user-attachments/assets/92901c21-a608-4b71-a5de-5d2114f6f5c6)

**Reply by colmjude** [reply]
I've created [a schema for the UPRN dataset](https://github.com/digital-land/specification/blob/main/content/dataset/uprn.md?plain=1).

I've not added an entity range yet whilst we decided how to handle this dataset.

Things to decide:

* what should the entity range be?
* the dataset from OS only includes point data, in this scenario do we need to include a `geometry` field in the dataset? We normally pair `point` and `geometry`

**Comment by eveleighoj** [comment]
As an idea on how we. can bring this data in. From the above we know that UPRNs are a well kept dataset and they have their own unique identifiers, Instead of assigning lookups could we calculate and add an entity column as the `entity = UPRN + X` where X is an offset to get the numbers inline with whatever entity range we decide internally. then we don't have to manage the lookups?

Also just a question on what happens when they retire a UPRN? Does this happen? Is it recorded somewhere in this dataset? Do we need to get information from another dataset?

**Reply by psd** [reply]
Moving from configuration to a rule to map references to entities is something we've discussed a few times. Some thoughts

1. We'd need need to create a large range to cover UPRNs if the mapping is as simple as `+ X` 
2. We could move entity numbers to be strings and add a short prefix, say "U" for UPRNs "I" for Inspire polygons, etc. 
3. As well as lookup we'd need similar rule for the entity-organisation and to have benefit we'd need to look at how we reduce the size of the facts table  
4. How many large datasets would this work for? (we have UPRNs, Index Polygons, Planning applications and decisions. What else is likely to benefit from such a rule?)
5. This feels like trading complexity to reduce the size of data. Is the size of the config a problem? We could start with code to generate the lookups before replacing the lookup, organisation and fact code with APIs.

**Reply by colmjude** [reply]
USRNs and EPCs are two more large datasets where this may work.

EPC references are like this `9768-5002-7205-6941-5960`

**Reply by psd** [reply]
On the need for Historic UPRNs, there is some information in this FOI request: https://www.whatdotheyknow.com/request/lists_of_historic_and_parent_upr

**Comment by PRWhitehouse** [comment]
There are a number of points to make on UPRNs (I was involved in there birth). BS7666 is the rule book. Main points for now are: the UPRN is the key to many dataset now, so adding prefixes will the pure link; speak to GeoPlace about use and access; note few local authorises have created a UPRN for all buildings or even land parcels, can be an issue for planning.
